src/libexpr/eval.cc

```diff
@@ -246,15 +246,37 @@ EvalState::EvalState(
     , repair(NoRepair)
     , emptyBindings(0)
     , rootFS(
-        settings.restrictEval || settings.pureEval
-        ? ref<SourceAccessor>(AllowListSourceAccessor::create(getFSSourceAccessor(), {},
-            [&settings](const CanonPath & path) -> RestrictedPathError {
-                auto modeInformation = settings.pureEval
-                    ? "in pure evaluation mode (use '--impure' to override)"
-                    : "in restricted mode";
-                throw RestrictedPathError("access to absolute path '%1%' is forbidden %2%", path, modeInformation);
-            }))
-        : getFSSourceAccessor())
+        ({
+            auto accessor = getFSSourceAccessor();
+
+            /* If we have a chroot store, make a union accessor to
+               make the chroot store available at its logical location
+               while still having the underlying directory
+               available. This is necessary for instance if we're
+               evaluating a file from the physical /nix/store while
+               using a chroot store. */
+            auto realStoreDir = dirOf(store->toRealPath(StorePath::dummy));
+            if (store->storeDir != realStoreDir) {
+                auto storeFS = makeMountedSourceAccessor(
+                    {
+                        {CanonPath::root, makeEmptySourceAccessor()},
+                        {CanonPath(store->storeDir), makeFSSourceAccessor(realStoreDir)}
+                    });
+                accessor = makeUnionSourceAccessor({accessor, storeFS});
+            }
```

Comment:
Why can we not use `getFSSourceAccessor` for this? Also I want `getFSSourceAccessor` to be rooted int the store dir. I don't like all these "union in place" things that both this and `getFSSourceAccessor` do. I want to do `CanonPath(storePath.to_string())` and have that just work with `storeFS` (note that the canon path will be e.g. `w4l4xvw461ywc4ia3accj5i3hh50n4r8-nix-2.24.1` in this case.

Reply:
CC @puffnfresh who I have been talking to about this same stuff